File: api/schemas/market.py
"""
Market schemas - Grand Exchange style order book
"""
from pydantic import BaseModel, Field, field_serializer
from typing import Optional, List
from datetime import datetime
from enum import Enum


def serialize_datetime_with_z(dt: Optional[datetime]) -> Optional[str]:
    """Serialize datetime to ISO8601 string with Z suffix for iOS compatibility"""
    if dt is None:
        return None
    # Strip microseconds - Swift's .iso8601 decoder can't parse them
    dt_no_micro = dt.replace(microsecond=0)
    iso_str = dt_no_micro.isoformat()
    if iso_str.endswith('+00:00'):
        return iso_str.replace('+00:00', 'Z')
    elif not iso_str.endswith('Z') and '+' not in iso_str and '-' not in iso_str[-6:]:
        # Naive datetime - assume UTC and add Z
        return iso_str + 'Z'
    return iso_str


class OrderType(str, Enum):
    """Order type"""
    BUY = "buy"
    SELL = "sell"


class OrderStatus(str, Enum):
    """Order status"""
    ACTIVE = "active"
    FILLED = "filled"
    PARTIALLY_FILLED = "partially_filled"
    CANCELLED = "cancelled"


# ItemType is now a simple string validated against RESOURCES config
# No more hardcoded enum! Frontend fetches available items from /market/available-items
ItemType = str  # Validated against routers.resources.RESOURCES at runtime


class CreateOrderRequest(BaseModel):
    """Request to create a new market order"""
    order_type: OrderType
    item_type: ItemType
    price_per_unit: int = Field(gt=0, description="Gold per unit")
    quantity: int = Field(gt=0, description="Number of units")


class MarketOrderResponse(BaseModel):
    """Response for a market order"""
    id: int
    player_id: int
    kingdom_id: str
    order_type: OrderType
    item_type: ItemType
    price_per_unit: int
    quantity_remaining: int
    quantity_original: int
    status: OrderStatus
    created_at: datetime
    updated_at: datetime
    filled_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True
    
    @field_serializer('created_at', 'updated_at', 'filled_at')
    @classmethod
    def serialize_dt(cls, dt: Optional[datetime]) -> Optional[str]:
        return serialize_datetime_with_z(dt)


class MarketTransactionResponse(BaseModel):
    """Response for a completed transaction"""
    id: int
    kingdom_id: str
    item_type: ItemType
    buyer_id: int
    seller_id: int
    quantity: int
    price_per_unit: int
    total_gold: int
    created_at: datetime
    
    class Config:
        from_attributes = True
    
    @field_serializer('created_at')
    @classmethod
    def serialize_dt(cls, dt: Optional[datetime]) -> Optional[str]:
        return serialize_datetime_with_z(dt)


class OrderBookEntry(BaseModel):
    """Single entry in the order book"""
    price_per_unit: int
    total_quantity: int  # Sum of all orders at this price
    num_orders: int      # Number of orders at this price


class OrderBook(BaseModel):
    """Current order book for an item"""
    item_type: ItemType
    kingdom_id: str
    
    # Buy orders (highest price first)
    buy_orders: List[OrderBookEntry]
    
    # Sell orders (lowest price first)
    sell_orders: List[OrderBookEntry]
    
    # Best prices
    highest_buy_offer: Optional[int] = None
    lowest_sell_offer: Optional[int] = None
    
    # Spread
    spread: Optional[int] = None  # Difference between best buy and sell


class PriceHistoryEntry(BaseModel):
    """Price history data point"""
    timestamp: datetime
    price: int
    quantity: int
    
    @field_serializer('timestamp')
    @classmethod
    def serialize_dt(cls, dt: Optional[datetime]) -> Optional[str]:
        return serialize_datetime_with_z(dt)


class PriceHistory(BaseModel):
    """Price history for an item"""
    item_type: ItemType
    kingdom_id: str
    transactions: List[PriceHistoryEntry]
    
    # Statistics
    average_price: Optional[float] = None
    min_price: Optional[int] = None
    max_price: Optional[int] = None
    total_volume: int = 0


class CreateOrderResult(BaseModel):
    """Result of creating an order (may include instant matches)"""
    order_created: bool
    order: Optional[MarketOrderResponse] = None
    
    # Instant matching results
    instant_matches: List[MarketTransactionResponse] = []
    total_quantity_filled: int = 0
    total_gold_exchanged: int = 0
    
    # Final status
    fully_filled: bool
    partially_filled: bool
    quantity_remaining: int = 0


class CancelOrderResult(BaseModel):
    """Result of cancelling an order"""
    success: bool
    message: str
    order_id: int
    refunded_items: Optional[int] = None  # For sell orders
    refunded_gold: Optional[int] = None   # For buy orders


class PlayerOrdersResponse(BaseModel):
    """Player's active and recent orders"""
    active_orders: List[MarketOrderResponse]
    recent_filled: List[MarketOrderResponse]
    recent_transactions: List[MarketTransactionResponse]


class MarketInfoResponse(BaseModel):
    """General market information"""
    kingdom_id: str
    kingdom_name: str
    market_level: int
    
    # Market access - requires home kingdom OR Merchant tier 3+
    can_access_market: bool = True
    
    # Available items (based on kingdom buildings)
    available_items: List[str]  # List of item_ids from resources.RESOURCES
    message: Optional[str] = None  # Message to display when no access or no items
    
    # Player resources
    player_gold: int
    player_resources: dict  # {item_type: quantity}
    
    # Market activity
    total_active_orders: int
    total_transactions_24h: int


class AvailableItemsResponse(BaseModel):
    """List of items available for trading with full config"""
    items: List[dict]  # Each item has: id, display_name, icon, color, description, category

File: ios/KingdomApp/KingdomApp/KingdomApp/Models/Market.swift
import Foundation

// MARK: - Enums

enum OrderType: String, Codable, CaseIterable {
    case buy = "buy"
    case sell = "sell"
}

enum OrderStatus: String, Codable {
    case active = "active"
    case filled = "filled"
    case partiallyFilled = "partially_filled"
    case cancelled = "cancelled"
    
    var displayName: String {
        switch self {
        case .active: return "Active"
        case .filled: return "Filled"
        case .partiallyFilled: return "Partially Filled"
        case .cancelled: return "Cancelled"
        }
    }
}

// ItemType is now dynamic - fetched from /market/available-items
// No more hardcoded enum! Use MarketItem for display properties.
typealias ItemType = String

// MARK: - Dynamic Item Config (from /market/available-items)

struct MarketItem: Codable, Identifiable, Hashable {
    var id: String { itemId }
    let itemId: String
    let displayName: String
    let icon: String        // SF Symbol name
    let color: String       // SwiftUI color name
    let description: String
    let category: String
    
    enum CodingKeys: String, CodingKey {
        case itemId = "id"
        case displayName = "display_name"
        case icon
        case color
        case description
        case category
    }
}

struct AvailableItemsResponse: Codable {
    let items: [MarketItem]
}

// MARK: - Models

struct MarketOrder: Codable, Identifiable {
    let id: Int
    let playerId: Int
    let kingdomId: String
    let orderType: OrderType
    let itemType: ItemType
    let pricePerUnit: Int
    let quantityRemaining: Int
    let quantityOriginal: Int
    let status: OrderStatus
    let createdAt: Date
    let updatedAt: Date
    let filledAt: Date?
    
    enum CodingKeys: String, CodingKey {
        case id
        case playerId = "player_id"
        case kingdomId = "kingdom_id"
        case orderType = "order_type"
        case itemType = "item_type"
        case pricePerUnit = "price_per_unit"
        case quantityRemaining = "quantity_remaining"
        case quantityOriginal = "quantity_original"
        case status
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case filledAt = "filled_at"
    }
}

struct MarketTransaction: Codable, Identifiable {
    let id: Int
    let kingdomId: String
    let itemType: ItemType
    let buyerId: Int
    let sellerId: Int
    let quantity: Int
    let pricePerUnit: Int
    let totalGold: Int
    let createdAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id
        case kingdomId = "kingdom_id"
        case itemType = "item_type"
        case buyerId = "buyer_id"
        case sellerId = "seller_id"
        case quantity
        case pricePerUnit = "price_per_unit"
        case totalGold = "total_gold"
        case createdAt = "created_at"
    }
}

struct OrderBookEntry: Codable {
    let pricePerUnit: Int
    let totalQuantity: Int
    let numOrders: Int
    
    enum CodingKeys: String, CodingKey {
        case pricePerUnit = "price_per_unit"
        case totalQuantity = "total_quantity"
        case numOrders = "num_orders"
    }
}

struct OrderBook: Codable {
    let itemType: ItemType
    let kingdomId: String
    let buyOrders: [OrderBookEntry]
    let sellOrders: [OrderBookEntry]
    let highestBuyOffer: Int?
    let lowestSellOffer: Int?
    let spread: Int?
    
    enum CodingKeys: String, CodingKey {
        case itemType = "item_type"
        case kingdomId = "kingdom_id"
        case buyOrders = "buy_orders"
        case sellOrders = "sell_orders"
        case highestBuyOffer = "highest_buy_offer"
        case lowestSellOffer = "lowest_sell_offer"
        case spread
    }
}

struct PriceHistoryEntry: Codable, Identifiable {
    var id: String { "\(timestamp)-\(price)" }
    let timestamp: Date
    let price: Int
    let quantity: Int
}

struct PriceHistory: Codable {
    let itemType: ItemType
    let kingdomId: String
    let transactions: [PriceHistoryEntry]
    let averagePrice: Double?
    let minPrice: Int?
    let maxPrice: Int?
    let totalVolume: Int
    
    enum CodingKeys: String, CodingKey {
        case itemType = "item_type"
        case kingdomId = "kingdom_id"
        case transactions
        case averagePrice = "average_price"
        case minPrice = "min_price"
        case maxPrice = "max_price"
        case totalVolume = "total_volume"
    }
}

struct CreateOrderResult: Codable {
    let orderCreated: Bool
    let order: MarketOrder?
    let instantMatches: [MarketTransaction]
    let totalQuantityFilled: Int
    let totalGoldExchanged: Int
    let fullyFilled: Bool
    let partiallyFilled: Bool
    let quantityRemaining: Int
    
    enum CodingKeys: String, CodingKey {
        case orderCreated = "order_created"
        case order
        case instantMatches = "instant_matches"
        case totalQuantityFilled = "total_quantity_filled"
        case totalGoldExchanged = "total_gold_exchanged"
        case fullyFilled = "fully_filled"
        case partiallyFilled = "partially_filled"
        case quantityRemaining = "quantity_remaining"
    }
}

struct PlayerOrdersResponse: Codable {
    let activeOrders: [MarketOrder]
    let recentFilled: [MarketOrder]
    let recentTransactions: [MarketTransaction]
    
    enum CodingKeys: String, CodingKey {
        case activeOrders = "active_orders"
        case recentFilled = "recent_filled"
        case recentTransactions = "recent_transactions"
    }
}

struct MarketInfo: Codable {
    let kingdomId: String
    let kingdomName: String
    let marketLevel: Int
    let canAccessMarket: Bool  // Requires home kingdom OR Merchant tier 3+
    let availableItems: [ItemType]
    let message: String?
    let playerGold: Int
    let playerResources: [String: Int]
    let totalActiveOrders: Int
    let totalTransactions24h: Int
    
    enum CodingKeys: String, CodingKey {
        case kingdomId = "kingdom_id"
        case kingdomName = "kingdom_name"
        case marketLevel = "market_level"
        case canAccessMarket = "can_access_market"
        case availableItems = "available_items"
        case message
        case playerGold = "player_gold"
        case playerResources = "player_resources"
        case totalActiveOrders = "total_active_orders"
        case totalTransactions24h = "total_transactions_24h"
    }
}

// MARK: - Request Bodies

struct CreateOrderRequest: Codable {
    let orderType: OrderType
    let itemType: ItemType
    let pricePerUnit: Int
    let quantity: Int
    
    enum CodingKeys: String, CodingKey {
        case orderType = "order_type"
        case itemType = "item_type"
        case pricePerUnit = "price_per_unit"
        case quantity
    }
}

